scalike-jdbc-scala-fx/src/main/scala/fx/DatabaseScala.scala

```diff
@@ -0,0 +1,22 @@
+package fx
+
+import scalikejdbc.*
+import scalikejdbc.DB.{readOnly, CPContext, NoCPContext}
+
+import java.sql.SQLException
+import fx.ensure
+
+type Database[A] = Control[SQLException] ?=> A
+type Transaction[A] = Control[Exception] ?=> A
+
+extension (db: DB.type)
+  def readOnlyWithControl[A](execution: DBSession => A)(
```

Comment:
@pablogthuillier I wonder, could we get a referential transparency test for these methods? I don't think they are eager outside of a `run` context, but `testFX` eagerly applies the run so it is unclear. Anything that shows that subsequent calls to a `Database[A]` evaluate every time. What I'm getting at is that if it isn't RT, the `A` returns may have to be `() => A` or `Function0[A]` instead of just `A`.

Basically in:

```scala
val names1 = names // from the test
val names2 = names
```

names should evaluate twice. You could do that by updating a `var` counter or whatever. `Transaction` is easier to test because you can read afterwards to see that two items are inserted.